---
title: "20230725_TCGA_OV_stratification"
output: html_document
date: "2023-07-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(janitor)
library(here)
library(readxl)
```

#Data required:
- *Y:/ecotyper/20230720_TCGA_2nd_year/GDCdata/20230815_clinical.csv* clinical data from GDC download produced in temp_GDC_download_serverversion.Rmd (jblab server002)
- additional datasets from:
-Wang, Y. K. et al. Genomic consequences of aberrant DNA repair mechanisms stratify ovarian cancer histotypes. Nat. Genet. 49, 856–864 (2017).
-Maxwell, K. N. et al. BRCA locus-specific loss of heterozygosity in germline BRCA1 and BRCA2 carriers. Nat. Commun. 8, 319 (2017).
-Drews, R.M., Hernando, B., Tarabichi, M. et al. A pan-cancer compendium of chromosomal instability. Nature 606, 976–983 (2022). 
-Cancer Genome Atlas Research Network. Integrated genomic analyses of ovarian carcinoma. Nature 474, 609–615 (2011).
-Takaya, H., Nakai, H., Takamatsu, S. et al. Homologous recombination deficiency status-based classification of high-grade serous ovarian carcinoma. Sci Rep 10, 2757 (2020).
-Shah, J.B., Pueschl, D., Wubbenhorst, B. et al. Analysis of matched primary and recurrent BRCA1/2 mutation-associated tumors identifies recurrence-specific drivers. Nat Commun 13, 6728 (2022).
- location "I:/research/jblab/data/group_folders/emily_lythgoe/20230724_EcoTyper_2nd_year/Data/Raw"

- *20230815_rnaseq_cases_missing_from_pancan.csv* produced in "20230815_DGE.Rmd"

```{r TCGA 2011 paper}
#specify data location using here() function
#data <- read_csv(here("./datafolder/subfolder/datafile.csv"))

#load BRCA1/2 data from Integrated Genomic Analyses of Ovarian Carcinoma (TCGA., 2011)
gBRCA1_cases<- read_excel(here("Data/Raw/2011_Integrated_genomic_analyses_of_Ov_TCGA/20230724_BRCA_data_from_Integrated_Genomic_Analyses_of_Ov_TCGA_2011.xlsx"), 
    sheet = "BRCA1_germline_mut") %>% 
   row_to_names(row_number = 2) %>% 
  clean_names() 

gBRCA2_cases<- read_excel(here("Data/Raw/2011_Integrated_genomic_analyses_of_Ov_TCGA/20230724_BRCA_data_from_Integrated_Genomic_Analyses_of_Ov_TCGA_2011.xlsx"), 
    sheet = "BRCA2_germline_mut") %>% 
   row_to_names(row_number = 2) %>% 
  clean_names() 

sBRCA1_cases<- read_excel(here("Data/Raw/2011_Integrated_genomic_analyses_of_Ov_TCGA/20230724_BRCA_data_from_Integrated_Genomic_Analyses_of_Ov_TCGA_2011.xlsx"), 
    sheet = "BRCA1_somatic_mut") %>% 
   row_to_names(row_number = 2) %>% 
  clean_names() 

sBRCA2_cases<- read_excel(here("Data/Raw/2011_Integrated_genomic_analyses_of_Ov_TCGA/20230724_BRCA_data_from_Integrated_Genomic_Analyses_of_Ov_TCGA_2011.xlsx"), 
    sheet = "BRCA2_somatic_mut") %>% 
   row_to_names(row_number = 2) %>% 
  clean_names() 

TCGA_all <- full_join(gBRCA1_cases, gBRCA2_cases) %>% 
  full_join(sBRCA1_cases) %>% 
  full_join(sBRCA2_cases)
```

```{r Pancancer paper}

#load supplementary files used in pancan paper to classify cases
BRCA_Wang_2017_all <- read_excel(here("Data/Raw/2022_A_pan-cancer_compendium_of_CIN/2017_Genomic_consequences_Wang/41588_2017_BFng3849_MOESM366_ESM.xlsx")) %>% 
  clean_names() 

BRCA_loss_Wang_2017 <- BRCA_Wang_2017_all %>% 
  dplyr::filter(brca_status!="WT") 

germline_mut_type_Maxwell_2017 <- read_excel(here("Data/Raw/2022_A_pan-cancer_compendium_of_CIN/2017_BRCA_locusspecific_LOH_Maxwell/41467_2017_388_MOESM2_ESM.xlsx")) %>% 
  row_to_names(row_number = 1) %>% 
  clean_names()  %>% 
  dplyr::filter(group=="TCGA",
                            tumor_site=="Ovary") %>% 
  as.data.frame()

#load pancan classification data
pancan_annotation <- read_excel(here("Data/Raw/2022_A_pan-cancer_compendium_of_CIN/41586_2022_4789_MOESM5_ESM/Supplementary_Tables_58-64.xlsx"), 
    sheet = "ST_58_Clinical_data_TCGA-OV") %>% 
  clean_names() %>% 
  group_by(status)

```


```{r}
tabyl(pancan_annotation$status) # numbers differ to the numbers reported in the supplementary methods file, missing 10 germline BRCA2 + LOH cases and 10 BRCA1 hypermethylation (BRCA1 promoter) cases and have gained 20 WT BRCA1/2+LOH (miscategorised)
```

```{r identify the missing BRCA1 promoter hypermethylation cases} 

#list of low BRCA1 mRNA expression cases from cBioPortal
BRCA1_lowexp_cBioPortal <- read_delim(here("Data/Raw/20230726_BRCA1_lowexp_cBioPortal.txt"), 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE) %>%  
  clean_names() #append = FALSE to write over existing file, 

 BRCA1_lowexp_cBioPortal <- BRCA1_lowexp_cBioPortal %>%  dplyr::rename( name = `patient_id` )


cBio_pancan_low_BRCA1 <- semi_join(pancan_annotation, BRCA1_lowexp_cBioPortal, by = "name")

   tabyl(cBio_pancan_low_BRCA1$status)

potential_hypermeth <- cBio_pancan_low_BRCA1 %>% 
  dplyr::filter(status == "WT BRCA1/2+LOH") # 8 samples with low BRCA1 expression in the WT+LOH category
#each of these samples show high levels of BRCA1 promoter methylation in cBioPortal so assume they are BRCA1 low expression due to hypermethylation and transfer to BRCA1/2 loss group 

#change status
missing_hypermeth <- potential_hypermeth %>% 
 mutate(status = replace(status, status == "WT BRCA1/2+LOH", "BRCA1 Hypermethyl.")) #conditional mutate


#pull out BRCA1/2 loss cases 
pancan_BRCA_LOF <- pancan_annotation %>% 
 dplyr::filter(status!="WT BRCA1/2",
         status!="WT BRCA1/2+LOH",
         status!="RAD51C Hypermethyl.")  


```




```{r, identifying missing gBRCA2+LOH cases}
gBRCA2_Wang <- BRCA_loss_Wang_2017 %>%  
  dplyr::filter(brca_status == "BRCA2.Germline") %>% 
  dplyr::rename(name = case_id)

#TCGA-24-1562, TCGA-13-1498, TCGA-24-2293, TCGA-24-1463 are gBRCA2 but not LOH according to the TCGA 2011 paper 

potential_gBRCA2LOH <- gBRCA2_Wang %>% 
  subset(name!="TCGA-24-1562" & 
         name!="TCGA-13-1498" & 
         name!="TCGA-24-2293" & 
         name!="TCGA-24-1463")

missing_gBRCA2LOH <- anti_join(potential_gBRCA2LOH, pancan_BRCA_LOF, by = "name") # only TCGA-13-0913 is missing from BRCA loss group...it is present, just with an a on the end TCGA-13-0913-a

#look in TCGA 2011 paper for additional gBRCA2 cases
gBRCA2_cases <- gBRCA2_cases %>% 
  dplyr::rename(name = case_id)

potential_gBRCA2LOH <- anti_join(gBRCA2_cases, pancan_annotation, by = "name") #  TCGA-24-2293 not present in the pancancer list at all

potential_gBRCA2LOH <- anti_join(pancan_annotation, gBRCA2_cases, by = "name")  
potential_gBRCA2LOH <-  anti_join(pancan_annotation, potential_gBRCA2LOH, by = "name") # 	in TCGA 2011 "TCGA-13-0900" is gBRCA2 + LOH,  TCGA-23-1026, TCGA-23-1026 and  TCGA-13-1512 are gBRCA2 but diploid, check likely impact of their mutations
#TCGA-13-1512 is a double mutant with gBRCA1+LOH and gBRCA2- keep in the original gBRCA1+LOH group
```

```{r}
#look in Maxwell 2017

germline_maxwell_brca2 <-   separate(germline_mut_type_Maxwell_2017, mutation, into = c("brca_status", "mutation"),  sep = " ") %>%  
  dplyr::filter(brca_status == "BRCA2") %>% 
  dplyr::rename(name = tumor_id) # some have two spaces as delimiter so the mutation info has been discarded in these cases but is in the original table

potential_gBRCA2LOH <- anti_join(germline_maxwell_brca2, pancan_BRCA_LOF) #12 samples with germline BRCA2 mutations that aren't listed as BRCA2 in pancancer annotation
#check if present in the pancancer dataset 
missing_gBRCA2LOH <- subset(pancan_annotation, name %in% c("TCGA-13-0751",
                                      "TCGA-13-0913",
                                      "TCGA-20-1685",
                                      "TCGA-23-1021",
                                      "TCGA-23-1029",
                                      "TCGA-23-1114",
                                      "TCGA-23-2084",
                                      "TCGA-24-1849",
                                      "TCGA-24-2293",
                                      "TCGA-25-1314",
                                      "TCGA-29-1784",
                                      "TCGA-59-A5PD"))  %>% 
  mutate(status = replace(status, status == "WT BRCA1/2+LOH", "germline BRCA2+LOH")) # 10 of the samples are present but listed as WT + LOH
#change status

#adding in hypermethylated and gBRCA2+LOH cases
pancan_BRCA_LOF <- full_join(pancan_BRCA_LOF, missing_hypermeth) 
pancan_BRCA_LOF <- full_join(pancan_BRCA_LOF, missing_gBRCA2LOH) %>% 
  mutate(contrast_group = "BRCA1_2_loss")


```
```{r check numbers}
tabyl(pancan_BRCA_LOF$status) 

tabyl(pancan_BRCA_LOF$myriadmy_choice) # 5 have a negative myriad score, could move these into the other group but I think BRCA1/2 status outranks myriad score clinically

```

```{r}
#the rest of cases
pancan_rest <- pancan_annotation %>% 
  anti_join(pancan_BRCA_LOF, by = "name") 

#separate out by myriad score
pancan_WT_HRD <- pancan_rest %>% 
  dplyr::filter(myriadmy_choice == "Positive") %>% 
  mutate(contrast_group = "HRD_high")

pancan_WT_HRP <- pancan_rest %>% 
  dplyr::filter(myriadmy_choice == "Negative") %>% 
  mutate(contrast_group = "HRD_low")

pancan_NA <- pancan_rest %>%
  dplyr::filter(myriadmy_choice == "NA") %>% 
  mutate(contrast_group = "NA") # not sure where to put these - assume HRP or disregard (check which actually have mRNA data)
```


```{r}
pancan_contrast <- full_join(pancan_BRCA_LOF, pancan_WT_HRD) %>% 
  full_join(pancan_WT_HRP) %>% 
  full_join(pancan_NA)

tabyl(pancan_contrast$status)
```
```{r}
#save a copy of the pancancer clinical data with the contrast labels added
 write_delim(pancan_contrast, "./Data/Processed/20230815_ST_58_Clinical_data_TCGA-OV_contrast_groups.tsv", delim = "\t") # save with contrast groups added
```


#import list of TCGA OV cases with RNAseq data that don't match a label in the pancancer dataset

```{r}
cases_missing_labels <- read_delim("Data/Processed/20230815_rnaseq_cases_missing_from_pancan.tsv", delim = "\t") %>% 
  dplyr::rename(case_id = submitter_id)
```

```{r}
missing_search_1 <- semi_join( BRCA_Wang_2017_all,cases_missing_labels) #17 cases present in the BRCA_Wang dataset

HRD_survival_2020 <- readxl::read_excel("Data/Raw/2020_Homologous recombination deficiency status-based classification of high-grade serous ovarian carcinoma/20230725_supp_ov_table_extracted_HRD_survival.xlsx", 
    sheet = "Table002 (Page 7)") %>% 
  row_to_names(row_number = 1) %>% 
  clean_names() %>% 
  dplyr::rename(case_id = patient_id)

missing_search_2 <- semi_join(HRD_survival_2020, cases_missing_labels) 

missing_search_3 <- semi_join(TCGA_all, cases_missing_labels) # only 1 case present in the TCGA 2011 BRCA1/2 loss annotation TCGA-24-2293, a gBRCA2 case

```
```{r}
#reload the cases_missing_labels list annotated using Wang 2017 and Takaya 2020 annotation
cases_missing_labels <- read_csv("Data/Processed/20230815_rnaseq_cases_missing_from_pancan.csv") %>% 
  dplyr::rename(tcga_tumor = submitter_id)

#HRD classification from Shah 2022
gBRCA_Shah_2022 <- read_excel("Data/Raw/2022_Analysis_of_matched_primary_and_recurrent_BRCA12/41467_2022_34523_MOESM14_ESM.xlsx", sheet = "TCGAOvarian_gBRCA") %>% 
  clean_names() 

gBRCA_Shah_2022$tcga_tumour <- gBRCA_Shah_2022$tcga_tumor %>%  
  str_trunc(12, ellipsis = "") # shorten name 

WT_Shah_2022 <- read_excel("Data/Raw/2022_Analysis_of_matched_primary_and_recurrent_BRCA12/41467_2022_34523_MOESM14_ESM.xlsx", sheet = "TCGAOvarian_HR-WT") %>% 
  clean_names()

WT_Shah_2022$tcga_tumor <- WT_Shah_2022$tcga_tumor %>% 
  str_trunc(12, ellipsis = "")

# TCGA-25-1328 listed as WT in Shah 2022, other than that, doesn't provide additional labels

```

```{r}
left_join(cases_missing_labels, gBRCA_Shah_2022)
```

