---
title: "DGE"
output: html_document
date: "2023-08-15"
---


```{r, setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

- based on DGE script from Ashley Weir 

**Data required:**
- **20230815_STARcounts_TCGAOV.tsv** produced in temp_GDC_download_serverversion.Rmd (original on jblab server2, put copies on I drive)
- Meta data from TCGA download **20230815_clinical.csv** (original on jblab server2, put copies on I drive)
- Location on jblab server002: /home/lythgo02/ecotyper/20230720_TCGA_2nd_year/GDCdata/ 
- Contrast labels **20230815_ST_58_Clinical_data_TCGA-OV_contrast_groups.tsv** produced in 20230725_TCGA_OV_stratification.Rmd 
- Location on I-drive: ./Data/Processed/


```{r, message=FALSE, warning=FALSE}
setwd("/mnt/jblab/group_folders/emily_lythgoe/20230724_EcoTyper_2nd_year")

library(dplyr)
library(tidyverse) #will need to install dependencies via terminal first: sudo apt install libcurl4-openssl-dev libssl-dev libxml2-dev libfontconfig1-dev libharfbuzz-dev libfribidi-dev libfreetype6-dev libpng-dev libtiff5-dev libjpeg-dev
library(ggplot2)
library(stringr)
library(AnnotationDbi) # install from CRAN
library(org.Hs.eg.db) #install from CRAN
library(limma) #install from CRAN
library(edgeR)
library(DESeq2)
library(janitor)
library(statmod)
library(GO.db)
```

**Load data**

```{r, message=FALSE}

rawCounts <- read_delim("/mnt/jblab/group_folders/emily_lythgoe/20230724_EcoTyper_2nd_year/Data/GDCdata/TCGA-OV/20230815_STARcounts_TCGAOV.tsv", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)

meta <- read_csv("/mnt/jblab/group_folders/emily_lythgoe/20230724_EcoTyper_2nd_year/Data/GDCdata/TCGA-OV/20230815_clinical.csv") 

#need to check for the most recent version of the pan-cancer clinical data table (modified to contrast contrast labels in Rmd 20230725_TCGA_OV_stratification.Rmd on I drive
pancan_contrasts <-  read_delim("Data/Processed/20230815_ST_58_Clinical_data_TCGA-OV_contrast_groups.tsv", 
                        delim = "\t", escape_double = FALSE, trim_ws = TRUE) %>% 
                        dplyr::rename(submitter_id = name) %>%
                        mutate(submitter_id = str_trunc(submitter_id, 12, "right", ellipsis = "")) # some samples have additional letters at the end (unsure if this is vial number), remove 

#CE information I generated using TPM processed by Tatlow and Piccolo,. 2016
ecotype_abundances <- read_delim("/mnt/jblab/group_folders/emily_lythgoe/20230724_EcoTyper_2nd_year/Output/Data/20230921_EcoTyperResults_TatlowPiccoloTPMmedian/Carcinoma_Ecotypes/Ecotype_Assignment.txt", delim = "\t") %>% 
  as.data.frame(transpose()) %>% 
  clean_names()

#CE information produced by Luca etal., 2021, only has info for 266 samples 
#ecotype_abundances <- read_delim("Data/Processed/20230920_Luca_CEdataOV.tsv", delim = "\t") %>% 
     #dplyr::rename(submitter_id = aliquot_barcode)  %>% 
     #dplyr::select(-cg_hub_analysis_id, -aliquot_id, -disease)
```

**Wrangle data**
Filter to include only primary samples. Some individuals have multiple sequencing files eg TCGA-61-2008-01A... and TCGA-61-2008-02A... the 7-8th number in the sequence indicates the sample type, 01 is a primary tumour and 02 is a recurring tumour (see GDC sample type codes).

```{r wrangle data,warning=FALSE, message=FALSE}

rawCounts <- rawCounts %>% 
  column_to_rownames(var = "gene_id") #430 cases with RNAseq counts from GDC
rawCounts <- rawCounts %>% 
  dplyr::select(-matches("02A|TCGA-23-1023-01R")) # remove any secondary samples and TCGA-23-1023-01R (have TCGA-23-1023-01A in addition which I have kept)

#the gene names are already present within the rawCounts data - use the annotation package instead as this is more reliable for cross-matching

primary_samples <- as.data.frame(colnames(rawCounts)) %>% 
   str_trunc(16, "right", ellipsis = "") %>% 
  clean_names() %>% 
  dplyr::rename(submitter_id = colnames_raw_counts) %>% 
  str_trunc(12, "right", ellipsis = "") 
```

```{r, eval=FALSE, include = FALSE}
#filtered_pancan <- left_join(primary_samples, pancan_contrasts) 

# pull out cases not present in the pancancer table
#cases_missing_labels <- filtered_pancan %>% 
 # filter(is.na(cx3)) %>% 
 # dplyr::select(submitter_id) %>% 
 # write_delim("./Data/Processed/20230815_rnaseq_cases_missing_from_pancan.tsv", delim = "\t")

```

```{r, message=FALSE}
#reload the cases_missing_labels list annotated using Wang 2017 and Takaya 2020 annotation
cases_missing_labels <- read_csv("Data/Processed/20230815_rnaseq_cases_missing_from_pancan.csv")

pancan_contrasts_shortened <- pancan_contrasts %>% 
  dplyr::select("submitter_id", "contrast_group") #only need the contrast info for now
 
meta_plus_contrast <- cases_missing_labels %>% 
  dplyr::select("submitter_id", "contrast_group") %>% 
  full_join(meta)

# add in the contrast group info from cases_missing_labels into meta_plus_contrast group
meta_plus_contrast <- meta_plus_contrast %>%  
  left_join(pancan_contrasts_shortened, by = ("submitter_id")) %>%  
  mutate(contrast_group = if_else(is.na(contrast_group.y), contrast_group.x, contrast_group.y)) %>% 
  dplyr::select(-contrast_group.x, -contrast_group.y) 
```

```{r, notes to self, include=FALSE, echo=FALSE}
#mutate line: the left_join merges the two tables by ID but keeps both contrast columns (renamed as contrast_group.y and contrast_group.x). The mutate creates a new column contrast_group, checks for NA in the contrast_group.y column and enters the value of the contrast_group.x column in the new contrast_group column if it is NA, otherwise it enters the value in the conrast_group.y column. 

#note if using if_else() from dplyr, use | and & as comparison operators, but if using if and else from base r, use || and && as comparison operators
```

```{r, wrangle ecotype data - Luca CE output, include=FALSE, eval=FALSE}

# <- ecotype_abundances %>% 
  #mutate(submitter_id_trunc = str_trunc(submitter_id, 12, "right", ellipsis = ""))
  
#subset_rows <- ecotype_abundances %>%
 # filter(duplicated(submitter_id_trunc) | duplicated(submitter_id_trunc, fromLast = TRUE)) ##The first call to duplicated marks the first occurrence of each duplicate, and the second call to duplicated with the fromLast = TRUE argument marks the last occurrence of each duplicate.

#ecotype_abundances <- ecotype_abundances %>%
 #dplyr::filter(!str_detect( submitter_id, "02A")) %>%  # discarding secondary samples
 # dplyr::select(-submitter_id) %>% 
  #dplyr::rename(submitter_id = submitter_id_trunc)

#meta_plus_contrast <- full_join( meta_plus_contrast, ecotype_abundances, by="submitter_id")
```

```{r, wrangle ecotype data - my CE output}

ecotype_abundances <- ecotype_abundances %>% 
 mutate(id=gsub("_", "-", toupper(id)), #replace underscores with hyphens and capitalise
        submitter_id_trunc = str_trunc(id, 12, "right", ellipsis = ""))
  
duplicate_rows <- ecotype_abundances %>%
  filter(duplicated(submitter_id_trunc) | duplicated(submitter_id_trunc, fromLast = TRUE)) ##The first call to duplicated marks the first occurrence of each duplicate, and the second call to duplicated with the fromLast = TRUE argument marks the last occurrence of each duplicate.

ecotype_abundances <- ecotype_abundances %>%
 dplyr::filter(!str_detect(id, "02A|TCGA-23-1023-01R-01R-1564-13-MEDIAN")) %>%  # discarding secondary and recurrance samples
  dplyr::select(-id) %>% 
  dplyr::rename(submitter_id = submitter_id_trunc)

meta_plus_contrast <- full_join( meta_plus_contrast, ecotype_abundances, by="submitter_id")


```


**Matching primary samples in rawcounts to samples in meta data**
```{r}
meta_rawCounts_matched <- meta_plus_contrast[meta_plus_contrast$submitter_id %in% primary_samples$submitter_id,]

merged_data <- primary_samples %>%
  left_join(meta_rawCounts_matched, by = c("submitter_id" = "submitter_id"))


meta_rawCounts_matched <- merged_data %>%
  arrange(match(submitter_id, primary_samples$submitter_id)) # Now ordered_data contains the rows from meta_rawCounts_matched in the same order as in primary_samples

colnames(rawCounts) <- substr(colnames(rawCounts), 1, 12)

missing_contrast <- meta_rawCounts_matched[is.na(meta_rawCounts_matched$contrast_group),]

rawCounts_final <- rawCounts[,!colnames(rawCounts) %in% missing_contrast$submitter_id] 
 rawCounts_final <- cbind(rawCounts$gene_name, rawCounts_final) %>% 
 dplyr::rename(gene_name = "rawCounts$gene_name")

meta_rawCounts_matched <- meta_rawCounts_matched[!is.na(meta_rawCounts_matched$contrast_group),]
```

**Annotate genes**

```{r geneAnnotation}
columns(org.Hs.eg.db)
colGene <- c("ENTREZID", "SYMBOL", "GENETYPE", "ENSEMBL") #defines the columns you want to select from the gene annotation database
ensembl <- data.frame(ensembl_id = rownames(rawCounts_final)) %>% 
  separate(ensembl_id, into = c("ensembl_id", "version_number"), sep="\\.") 


#query the org.Hs.eg.db gene annotation database using the Ensembl Gene IDs as keys, retrieving  the specified columns (colGene)
geneIdMap <- AnnotationDbi::select(org.Hs.eg.db, #the org.Hs.eg.db package provides annotations for the human genome
                                       keys = as.character(ensembl$ensembl_id), 
                                       columns = colGene, keytype = "ENSEMBL") 

count(is.na(geneIdMap$ENTREZID)) # 24622 don't match to an ID
duplicated_geneIdMap <- geneIdMap %>% 
  dplyr::filter(duplicated(ENTREZID) &!is.na(ENTREZID)) #2812 duplicated ID

geneIdMap <- geneIdMap[!duplicated(geneIdMap$ENTREZID), ]#remove duplicated rows from the geneIdMapExAn data frame based on the ENTREZID (retaining one copy of each duplicate)
geneIdMap <- geneIdMap[complete.cases(geneIdMap), ] # one gene has missing entries, remove

rawCounts_final <- rawCounts[geneIdMap$ENSEMBL, ]
rawCounts_final <- rawCounts_final[,!colnames(rawCounts) %in% missing_contrast$submitter_id]
rawCounts_final <- rawCounts_final[complete.cases(rawCounts_final), ]

ensembl <- data.frame(ensembl_id = rownames(rawCounts_final)) %>% 
  separate(ensembl_id, into = c("ensembl_id", "version_number"), sep="\\.") #regenerate filtered list of ENSEMBL to filter geneIdMap by
geneIdMap <- geneIdMap[geneIdMap$ENSEMBL %in% ensembl$ensembl_id, ] #filter to remove rows no longer present in rawCounts_final

```
**Alternate method of mapping**
**Using other method as this returns fewer mapped genes**
```{r}
# Load the AnnotationHub package
#library(AnnotationHub)

# Create an AnnotationHub object for Homo sapiens
#ah <- AnnotationHub()

# Create a list of Ensembl Gene IDs
#ensembl_ids <- rownames(rawCounts_final)
#ensembl_ids <- sub("\\..*", "", ensembl_ids)  # Remove version numbers

# Create a query to retrieve gene annotation data from AnnotationHub
#humanEnsDb <- query(ah, c("EnsDb", "Homo sapiens", "GRCh38"))[[1]]
#annotations <- genes(humanEnsDb, return.type = "data.frame")

#annot <- annotations %>%
  #dplyr::select(gene_id, gene_name, entrezid) %>%
  #dplyr::filter(gene_id %in% ensembl_ids)

#length(unique(annot$gene_id)) #23767

#sum(is.na(annot$entrezid)) # 11240 don't map to ID
```


```{r, Create rawCounts_final collapsed into median STAR count, include=FALSE, eval=FALSE}
#Not required as EcoTyper won't take STARcounts, only TMK/FPKM

#EcoTyper_counts <- rawCounts_final %>%
 # rownames_to_column("gene_id") %>%  
 # separate(gene_id, into = c("ENSEMBL", "version_number"), sep = "\\." ) %>% 
 # group_by(ENSEMBL) %>% 
 # summarize(across(where(is.numeric), median)) #collapse transcript variants into median value for that ENSEMBL ID as separating version numbers from ENSEMBL creates duplicates 

#EcoTyper_final <- EcoTyper_counts %>% 
  #inner_join(geneIdMap %>% dplyr::select(ENSEMBL, SYMBOL), by = "ENSEMBL")# %>% 
 #save <- EcoTyper_final %>%  dplyr::select(-ENSEMBL) %>% write_delim("./Data/Processed/20230908_EcoTyperCountsMappedMedian.tsv", delim="\t")

```


**Create a DGE object for DE analysis**

```{r, dgeObject}

dge <- DGEList(counts = rawCounts_final, genes = geneIdMap, samples = meta_rawCounts_matched, group = meta_rawCounts_matched$contrast_group) # contains all samples in one object 

```

**Filter lowly expressed genes.**
Using manual filtering and the filterByExp() function

```{r, filter low expression genes}

cutoff <- 4
drop <- which(apply(cpm(dge), 1, max) < cutoff) #cpm() calculates counts per million for normalisation, apply() is then used to calculate the max (using max()) CPM value across all rows (1 means all rows, 2 would mean all columns), and which() returns indices of genes where max cpm < 4
dgeFil <- dge[-drop, , keep.lib.sizes = FALSE] 
dim(dgeFil) # number of genes left after manual filtering

keep <- filterByExpr(dgeFil, group = dgeFil$samples$contrast_group) #filters genes with low counts 

dgeFilExp <- dgeFil[keep, ,keep.lib.sizes = FALSE ] 
dim(dgeFilExp) # number of genes left after manual filtering and filterByExpr

```

**Process all contrast groups together **
```{r}
#check for samples that have more counts on average than other samples 
boxplot(log(dgeFilExp$counts + 0.5), xaxt = "n", col = "turquoise4", xlab = "Samples") # raw log counts (+0.5 to prevent taking the log of 0)
boxplot(cpm(dgeFilExp$counts, log = TRUE), xaxt = "n", col = "turquoise4", xlab = "Samples") # log cpm normalised counts

plotSparsity(dgeFilExp$counts, normalized = TRUE) # normalised to account for variations in library size to avoid sequencing bias


```

**Normalisation**

```{r}

# calculate normalisation factors (does not normalise)
dgeFilExpNorm <- calcNormFactors(dgeFilExp)

barplot(dgeFilExpNorm$samples$norm.factors, las = 2)
title("Barplot of normalisation factors")

```

**Define groups for contrast interactions**
```{r}

 
group <-  interaction(dgeFilExpNorm$samples$contrast_group)
groupCol <- group # you could just use the original interaction vector, but I like to make a new one that I will overlay colours
levels(groupCol) <- c("#F8766D", "#00BFC4", "#00BA38") # change levels to colour names/HEX codes.
# BRCA1/2 Loss == turquoise (#00BFC4)
#HRD high == tomato ("#F8766D")
#HRD low == green ("#00BA38")
groupCol <- as.character(groupCol) 

plotMDS(dgeFilExpNorm, col = groupCol, labels = NULL, pch = 20) #multidimensional scaling plot to visualise similarity between samples based on gene expression profiles 
legend("topright", legend = c("BRCA1/2 loss", "HRD high", "HRD low"), fill =  c("#F8766D", "#00BFC4", "#00BA38"))
#distances on the plot approximate the typical log2 fold changes between the samples 

```
**defining covariate interactions**
May want to try collapsing the stages into 1-4, the separation may be too granular to see confounders at present 
```{r}

stage <- interaction(dgeFilExpNorm$samples$figo_stage)
stageCol <- stage
# Levels: Stage IC, IIA, IIB, IIC, IIIA, IIIB, IIIC, IV
levels(stageCol) <- c("darkorange", "cornflowerblue", "darkolivegreen3", "deeppink3", "darkorchid", "brown1", "grey", "chartreuse" ) 
stageCol <- as.character(stageCol) 

plotMDS(dgeFilExpNorm, col = stageCol, labels = NULL, pch = 20)
legend("topright", legend = c("Stage IC", "IIA", "IIB", "IIC", "IIIA", "IIIB", "IIIC", "IV"), fill =  c("darkorange", "cornflowerblue", "darkolivegreen3", "deeppink3", "darkorchid", "brown1", "grey", "chartreuse" ) )


```

```{r, EcoTyper assignments for dominant ecotype in each sample}

CE <- interaction(dgeFilExpNorm$samples$carcinoma_ecotype) #need to change to dominant_ecotype if using CE data from Luca et al instead of own
CEcol <- CE
levels(CEcol) <- c("darkorange", "cornflowerblue", "darkolivegreen1", "deeppink3", "darkorchid", "darkslateblue", "darkgrey", "chartreuse4", "darkgoldenrod4", "#00BFC4" , "black", "pink") 
CEcol <- as.character(CEcol) 

plotMDS(dgeFilExpNorm, col = CEcol, labels = NULL, pch = 20)
legend("topright", legend = c("CE1", "CE2", "CE3", "CE4", "CE5", "CE6", "CE7", "CE8", "CE9", "CE10", "Unassigned", "NA"), fill =  c("darkorange", "cornflowerblue", "darkolivegreen1", "deeppink3", "darkorchid", "darkslateblue", "darkgrey", "chartreuse4",  "darkgoldenrod4", "#00BFC4", "black", "pink" ) )


```
Some separation of ecotypes but not correcting for this as it is confounded with gene expression (used to generate the CE classification) and the contrasts 

```{r}

site <- interaction(dgeFilExpNorm$samples$site_of_resection_or_biopsy)
siteCol <- site
levels(siteCol) <- c("darkorange", "cornflowerblue")

plotMDS(dgeFilExpNorm, col = siteCol, labels = NULL, pch = 20)

```



**Define design matrix**

```{r}

modelMatrix <- model.matrix(~ 0 + group, data=dgeFilExpNorm) # specifies the design of the linear model

```

**Voom transformation** 

```{r}

voom <- voom(dgeFilExpNorm, modelMatrix, plot = TRUE) #transforms counts to log2 CPM based on the normalisation factors calculated earlier & fits a linear model to the log2 CPM for each gene (row-wise), calculates the residuals, and fits a smoothed curve to the square root of the residual standard deviation by average expression. The smoothed curve is used to obtain weights for each gene and sample, which are then passed into limma for DE, taking into account the variability of each gene.

```


**Fit linear model**

```{r}
# fitting a linear model using weighted least squares for each gene
fit <- lmFit(voom, modelMatrix) # now run DEA
head(coef(fit)) #display the coefficients for the model terms defined in the model matrix. These coefficients will be used to perform hypothesis testing to determine significantly up/down regulated genes. 

```

**Define contrast**
BRCA1/2 loss is the baseline for each contrast
```{r}
# comparison between groups (log-fold changes) are obtained as contrasts of these fitted linear models
contr <- makeContrasts( groupHRD_high - groupBRCA1_2_loss, groupHRD_low - groupBRCA1_2_loss, levels = modelMatrix) # define the comparisons you want to make
#  groupHRD_high - groupBRCA1_2_loss, # Coefficient 1
 # groupHRD_low - groupBRCA1_2_loss,  # Coefficient 2
contr
#-1 = reference
#0 = not included in comparison
#1 = group to compare against reference

contrastsfit <- contrasts.fit(fit, contrasts = contr) # fit the contrasts to the linear model fit and calculate the coefficients 
fittedContrasts <- eBayes(contrastsfit, robust = TRUE) # use robust emperical bayes for statitsical testing to account for potential issues with heteroscedasticy in the data. Helps to obtain more accurate p-values and effect size estimates.

```


**Extract to contrasts and statistical testing.**
The output from topTable() includes:
-logFC: the log fold-change between cases and controls
-t: the t-statistic for DE
-P.Value: for DE (not adjusted for multiple testing)
-adj.P.Val: p-value adjusted for multiple testing. Default adjustment method = Benjamini-Horchberg
```{r}

results_HRDhigh <- topTable(fittedContrasts, coef= "groupHRD_high - groupBRCA1_2_loss", sort.by = "P", n = Inf) 
head(results_HRDhigh, 20) # look at top 20 DE genes (sorted by p value), n=Inf means retreive all DE genes (no limit on number)

up_genes_BRCAloss_HRDhigh <- results_HRDhigh %>%
  dplyr::filter(adj.P.Val < 0.05, logFC >= 1.0) #%>% 
  write_csv("Output/Data/20230922_up_genes_BRCAloss_HRDhigh.csv")
  
down_genes_BRCAloss_HRDhigh <- results_HRDhigh %>%
  dplyr::filter(adj.P.Val < 0.05, logFC <= -1.0) #%>% 
  write_csv("Output/Data/20230922_down_genes_BRCAloss_HRDhigh.csv")

results_HRDlow <- topTable(fittedContrasts, coef= "groupHRD_low - groupBRCA1_2_loss", sort.by = "P", n = Inf)
head(results_HRDlow, 20)

up_genes_BRCAloss_HRDlow <- results_HRDlow %>%
  dplyr::filter(adj.P.Val < 0.05, logFC >= 1.0) #%>% 
   write_csv("Output/Data/20230926_up_genes_BRCAloss_HRDlow.csv") 

down_genes_BRCAloss_HRDlow <- results_HRDlow %>%
  dplyr::filter(adj.P.Val < 0.05, logFC <= -1.0) #%>% 
  write_csv("Output/Data/20230926_down_genes_BRCAloss_HRDlow.csv")



  venn_list <- list(
      up_in_HRD = as.character(up_genes_BRCAloss_HRDhigh$SYMBOL),
      up_in_HRP = as.character(up_genes_BRCAloss_HRDlow$SYMBOL),
      down_in_HRD = as.character(down_genes_BRCAloss_HRDhigh$SYMBOL),
      down_in_HRP = as.character(down_genes_BRCAloss_HRDlow$SYMBOL))
  
 
vennDat <- ggvenn::ggvenn(venn_list, set_name_size = 3)   

vennDat + labs(title = "In comparison to BRCA1/2 loss")

```

**Sanity checks**
```{r}
hist(results_HRDhigh$P.Value)

hist(results_HRDlow$P.Value)
```

**Just looking at BRCA1/2 vs HRD low**
```{r}

# Total number of DE genes based on adjusted p value < 0.05
length(which(results_HRDlow$adj.P.Val < 0.05))

# Total number of DE genes based on adjusted p value < 0.05 and |logFC| (absolute log fold-change) >= 1.0
length(which(results_HRDlow$adj.P.Val < 0.05 & abs(results_HRDlow$logFC) >= 1.0 ))

# Total number of DE genes based on adjusted p value < 0.01 and |logFC| >= 1.0
length(which(results_HRDlow$adj.P.Val < 0.01 & abs(results_HRDlow$logFC) >= 1.0 ))

dt <- decideTests(fittedContrasts, p.value = 0.05, lfc = 1.0) #makes DE gene calls based on a specified p-value and log-fold change threshold 
summary(dt)

```


**Plots for DE contrast: BRCA1/2 loss vs HRD low**
```{r}

#to plot a measure of significance against fold change 
volcanoplot(
  fittedContrasts,             
  coef = "groupHRD_low - groupBRCA1_2_loss", # Specify which contrast to visualize
  highlight = 10,               # The number of top genes to highlight 
  names = fittedContrasts$genes$SYMBOL,    
  main = "Volcano Plot"        # Plot title 
)

#x axis - upregulation (positive values) and downregulation (negative) of genes 
#y-axis - statistical significance 
#each gene = a dot

#MA plot displays log-fold change against the average log expression for each gene 
limma::plotMA(fittedContrasts, coef = "groupHRD_low - groupBRCA1_2_loss", status = dt[ ,"groupHRD_low - groupBRCA1_2_loss"], hl.pch = 20, bg.col = "grey") 
o <- order(fittedContrasts$p.value[,"groupHRD_low - groupBRCA1_2_loss"]) # orders the genes based on their p-value from DEA
x <- fittedContrasts$Amean #extracts the average log-expression values (Amean) for all genes 
y <- fittedContrasts$coefficients[,"groupHRD_low - groupBRCA1_2_loss"] #extracts logFCfor the contrast group 
G <- fittedContrasts$genes$SYMBOL #extracts the gene names for genes in the fittedContrasts objects 

x_offset <- 0.2  # Adjust the x-offset as needed
y_offset <- 0.1  # Adjust the y-offset as needed
text(x[o[1:10]] + x_offset, y[o[1:10]] + y_offset,labels = G[o[1:10]], # to show the top 10 DE gene symbols
     cex = 0.7,  # Adjust label size 
     col = "black",  # Change label color 
     srt = 45) 



```

```{r ggplot version}

o <- order(fittedContrasts$p.value[, "groupHRD_low - groupBRCA1_2_loss"])
x <- fittedContrasts$Amean
y <- fittedContrasts$coefficients[, "groupHRD_low - groupBRCA1_2_loss"]
G <- fittedContrasts$genes$SYMBOL

# Create the MA plot using ggplot2 with your existing fittedContrasts object
ggplot(data = fittedContrasts$genes, aes(x = x, y = y)) +
  geom_point(aes(color = fittedContrasts$p.value[, "groupHRD_low - groupBRCA1_2_loss"] < 0.05), size = 1) +
  geom_text(data = fittedContrasts$genes[o[1:10], ], aes(x = x[o[1:10]] + 0.2, y = y[o[1:10]] + 0.1, label = G[o[1:10]]), size = 3, color = "black", angle = 45) +
  labs(
    x = "Average Log-Expression",
    y = "Log-Fold Change",
    title = "MA Plot: HR Proficient vs BRCA1/2 Loss",
    color = "FDR < 0.05"
  ) +
  theme_minimal()

o2 <- order(fittedContrasts$p.value[, "groupHRD_high - groupBRCA1_2_loss"])
x <- fittedContrasts$Amean
y2 <- fittedContrasts$coefficients[, "groupHRD_high - groupBRCA1_2_loss"]
G <- fittedContrasts$genes$SYMBOL

ggplot(data = fittedContrasts$genes, aes(x = x, y = y2)) +
  geom_point(aes(color = fittedContrasts$p.value[, "groupHRD_high - groupBRCA1_2_loss"] < 0.05), size = 1) +
  geom_text(data = fittedContrasts$genes[o2[1:10], ], aes(x = x[o2[1:10]] + 0.2, y = y[o2[1:10]] + 0.1, label = G[o2[1:10]]), size = 3, color = "black", angle = 45) +
  labs(
    x = "Average Log-Expression",
    y = "Log-Fold Change",
    title = "MA Plot: Non-BRCA HRD vs BRCA1/2 Loss",
    color = "FDR < 0.05"
  ) 
```

```{r, include=FALSE, eval = FALSE}
library(pheatmap) 

# Select the top genes based on FDR (assuming you have FDR values)
topGenesHRDlow <- results_HRDlow %>% 
  arrange(adj.P.Val) %>% #ascending order
  head(300)  
  
logFC_matrix <- topGenesHRDlow %>% 
  dplyr::select(SYMBOL, logFC)

rownames(logFC_matrix) <- logFC_matrix[, "SYMBOL"]

logFC_matrix <- logFC_matrix %>% 
  dplyr::select(-"SYMBOL") %>% 
  as.matrix()  

pheatmap(logFC_matrix, 
         cluster_rows = TRUE,  # Hierarchical clustering for rows
         cluster_cols = TRUE, # No clustering for columns
         color = colorRampPalette(c("blue", "white", "red"))(50), # Color scheme
         main = "Heatmap of Top DE Genes", # Heatmap title
         fontsize_row = 2,      # Row label font size
         fontsize_col = 2,      # Column label font size
         show_rownames = TRUE,  # Show row names
         show_colnames = TRUE)  # Show column names







```
```{r}
library(clusterProfiler)

search_kegg_organism('human', by='common_name') #kegg code = hsa
```


**Gene ontology analysis with Goana**

```{r}
# subset DE genes to include those of great significance and large-ish logFC
genesDE_HRDhigh <- results_HRDhigh %>%
  mutate(bigLogFC = case_when(logFC < -1.0 | logFC > 1.0 ~ "keep",
                              TRUE ~ "discard")) %>% 
  dplyr::filter(bigLogFC == "keep") %>% 
  dplyr::filter(adj.P.Val < 0.05) %>% 
  dplyr::select(ENTREZID) 
genesDE_HRDhigh <- as.vector(genesDE_HRDhigh)  

## gene ontology
gene_ont <- goana(genesDE_HRDhigh)
 topGO(gene_ont)

gene_ont$adj.P.Val <- p.adjust(gene_ont$P.ENTREZID, method="BH")

gene_ont <- gene_ont %>% 
  rownames_to_column(var = "goTerm")

#write_csv(gene_ont, "Output/Data/TCGA_DGE_geneOntology_20230922.csv")

```

