---
title: "GDC downloads"
output: html_document
date: "2023-07-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

## R Markdown
Access and import TCGA data using the R/Bioconductor package TCGAbiolinks
#https://benbermanlab.com/assets/code/Workshop%20-%20TCGA%20data%20analysis.html

#http://www.ensembl.org/info/data/biomart/biomart_r_package.html

```{r, load required packages}

suppressMessages({
    library(TCGAbiolinks)
    library(MultiAssayExperiment)
    library(maftools)
    library(dplyr)
    library(ComplexHeatmap)
    library(readr)
    library(biomaRt)
  library(GenomicDataCommons)
  library(here)
})


```

GDC database connection:
In order to filter the data available in GDC some fields are available such as project (TCGA, TARGET, etc.), data category (Transcriptome Profiling, DNA methylation, Clinical, etc.), data type (Gene Expression Quantification, Isoform Expression Quantification, Methylation Beta Value, etc.), experimental strategy (miRNA-Seq, RNA-Seq, etc.), Workflow Type, platform, access type and others.

```{r, echo = FALSE}
clinical <- GDCquery_clinic("TCGA-OV") # see what is in the TCGA-OV project in GDC database
head(clinical)
```
Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
```{r}
query <- GDCquery(project = "TCGA-OV", 
                  data.category = "Transcriptome Profiling", 
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "STAR - Counts")
```
```{r download STAR Counts}
GDCdownload(query )# downloads individual files and manifest file, The downloaded data will be saved in a directory with the project name and a sub-folder with the data.category

#429 files 

```

```{r}
#searchable datatable summarising files available 
results <- getResults(query)

```

```{r}
#GDCprepare transforms the downloaded data into a summarizedExperiment object or a data frame
# Prepare expression matrix with geneID in the rows and samples (barcode) in the columns
counts <- GDCprepare(query) #%>% 
   # rename(ensembl_gene_id_version = X1) #rename to match ensemble biomart notation


```
Available assays in SummarizedExperiment : 
  => unstranded
  => stranded_first
  => stranded_second
  => tpm_unstrand
  => fpkm_unstrand
  => fpkm_uq_unstrand
  
#If a summarizedExperiment object was chosen, the data can be accessed with three different accessors: assay for the data information, rowRanges to gets the range of values in each row and colData to get the sample information (patient, batch, sample type, etc)

```{r}
library(SummarizedExperiment) 
# get expression matrix
 OVmatrix <- assay(counts, "unstranded") 
```

```{r}
# get genes information
genes.info <- as.data.frame(rowRanges(counts))  #table 

# get sample information
sample.info <- as.data.frame(colData(counts))
```

```{r}
genes.info.hgnc <- genes.info %>% dplyr::select(gene_id, gene_name)

OVtable <- OVmatrix %>% as.data.frame() %>% 
  rownames_to_column("gene_id") 

OVtable <- 
  left_join(OVtable, genes.info.hgnc) %>% 
  group_by(gene_name) #%>% 
  summarise(across(where(is.numeric),list(median = median)))

write_delim(OVtable, file="/home/lythgo02/ecotyper/20230720_TCGA_2nd_year/GDCdata/20230727_STARcounts_TCGAOV.tsv", delim="\t") # save count data 
write_csv(clinical, file="/home/lythgo02/ecotyper/20230720_TCGA_2nd_year/GDCdata/20230727_clinical.csv")
```

```{r}
dataPrep <- TCGAanalyze_Preprocessing( #performs array array intensity correlation to check for outliers
    object = counts, 
    cor.cut = 0.6,
    datatype = "unstranded"
)     


```
