---
title: "Adding annotation to ensemble IDs"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
https://bioinformatics-core-shared-training.github.io/Bulk_RNASeq_Course_Feb22/Bulk_RNAseq_Course_Base/Markdowns/S6_Annotation_With_BioMart.html
```{r}
library(biomaRt)
library(tidyverse)
```
There are a number of ways to add annotation. One method is to do this using the org.Mm.eg.db package. This package is one of several organism-level packages which are re-built every 6 months. These packages are listed on the annotation section of the Bioconductor, and are installed in the same way as regular Bioconductor packages.

An alternative approach is to use biomaRt, an interface to the BioMart resource. This is the method we will use today.

Select BioMart database and dataset
The first step is to select the Biomart database we are going to access and which data set we are going to use.

There are multiple mirror sites that we could use for access. The default is to use the European servers, however if the server is busy or inaccessible for some reason it is possible to access one of the three mirror sites. See the instructions at here for detailed instruction on using different mirrors, but in brief simply add the mirror argument to the listEnsembl and useEnsembl functions below.

e.g to use the US West mirror:
ensembl <- useEnsembl("genes", mirror = "uswest")

Pro Tip: The Ensembl servers tend to get very busy in the afternoons, to the extent that biomaRt may have trouble getting and maitaining a connection. Try to do this in the morning.

If the only annotation you can see is Ensembl Gene ID
```{r}
# view the available databases
listEnsembl()
```
We want the “genes” database.

The current Ensembl release is version 106, however, you need to use the version that was in use when you downloaded the files. We will need to ensure that biomaRt accesses the correct version of the gene annotations.
```{r}
# view the available databases
listEnsembl(version = 102)
```
Now we need to identify the dataset for the species we need.
```{r}
## set up connection to ensembl database
ensembl <- useEnsembl(biomart = "genes", version=102)

# serach the available datasets (species)
searchDatasets(mart = ensembl, pattern = "Human")
```

We want the hsapiens_gene_ensembl	dataset.

specify a data set to use
We can now specify the complete connection to the database and dataset that we want to access.
```{r}
ensembl <- useEnsembl(biomart = 'genes', 
                      dataset = 'hsapiens_gene_ensembl',
                      version = 102)
```

Query the database
Now we need to set up a query. For this we need to specify three things:

What type of information we are going to search the dataset on - called filters. In our case this is Ensembl Gene IDs
A vector of the values for our filter - the Ensembl Gene IDs from our DE results table
What columns (attributes) of the dataset we want returned.
```{r}
# check the available "filters" - things you can filter for
ensembl_filters  <- listFilters(ensembl)

# To find the correct name for the Ensembl ID we can filter the name column
ensembl_filters %>%  
        filter(str_detect(name, "ensembl"))
```

So, we will use ensembl_gene_id to query the data bases
```{r}

# check the available "attributes" - things you can retreive
ensembl_attributes <- listAttributes(ensembl)
head(ensembl_attributes, 20)

```

We’ll retrieve the external_gene_name, which is the Gene Symbol, the entrez_id - we may need this for tools that use the NCBI databases - and the entrez_accession, which is the Gene Symbol associated with that entrez_id. Also the hgnc_symbol, ensembl_transcript_id_version and ensembl_gene_id_version.

We also need to specify that we want the query to return the ensembl_gene_id that we used to query the database. This ensures that our final table will include all of query gene ids, even if no record was found in the data base.

Returning data from Biomart can take time, so it’s always a good idea to test your query on a small list of values first to make sure it is doing what you want. We’ll just use the first 1000 genes for now.

```{r}

# Set the filter type and values
ourFilterType <- "ensembl_gene_id_version"

# get the Ensembl IDs from our results table
filterValues <- TCGA_OV_tpm$`2` #giving it the ensembl gene ID ersion to filter for

# Set the list of attributes
attributeNames <- c("ensembl_gene_id",
                    "external_gene_name", 
                    "entrezgene_id", 
                    "entrezgene_accession",
                    "ensembl_transcript_id_version",
                    "ensembl_gene_id_version",
                     "hgnc_symbol")

# run the query
annot <- getBM(attributes=attributeNames, 
               filters = ourFilterType, 
               values = filterValues, 
               mart = ensembl)
head(annot)
```
Some genes have multiple entries in the retrieved annotation. This is because there are multiple Entrez IDs for a single Ensembl gene. These one-to-many relationships come up frequently in genomic databases, it is important to be aware of them and check when necessary.
```{r}
# find all rows containing duplicated ensembl ids
annot %>%  
    add_count(ensembl_gene_id) %>%  
    dplyr::filter(n>1) # need to specify dplyr because the GDC package masks the dypler filter function with it's own
```

We will need to do a little work before adding the annotation to our results table. We could decide to discard one or both of the Entrez ID mappings, or we could concatenate the Entrez IDs so that we don’t lose information. Matching the entrezgene_accession and the external_gene_name can help with resolving some of these problems.
```{r}

fixedDuplicates <- annot %>%  
    add_count(ensembl_gene_id) %>%  
    dplyr::filter(n>1) %>% 
    dplyr::select(-n) %>% 
    dplyr::filter(entrezgene_accession==external_gene_name)

annot <- annot %>%  
    add_count(ensembl_gene_id) %>%  
    dplyr::filter(n==1) %>% 
    dplyr::select(-n) %>% 
    bind_rows(fixedDuplicates)

nrow(annot)
```