---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, load required packages}

suppressMessages({
    library(TCGAbiolinks)
    library(MultiAssayExperiment)
    library(maftools)
    library(dplyr)
    library(ComplexHeatmap)
    library(readr)
    library(biomaRt)
  library(GenomicDataCommons)
})
```


```{r}
#load ecotyper reslts from analysis of whole cohort
ecotype_assignment <- read_delim("~/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcgca_all_samples_060722_ecotyper_output/hugo_tcga_all_samples_060722/Ecotypes/ecotype_assignment.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE) %>% 
  clean_names() %>% 
  dplyr::rename(assigned_ce=assignment_p)

#convert ID to same format as in hugo_brca table
ecotype_assignment$id <- str_replace_all(ecotype_assignment$id, "\\.", "-")
ecotype_assignment$id <- str_trunc(ecotype_assignment$id, 15)
ecotype_assignment$id <- str_replace_all(ecotype_assignment$id, "\\...", "")
```

```{r}
#use Ruben's analysis from "Pancancer compendium of chromosomal instability" Nature 2022 to classify cases as HRD negative or positive based on Myriad mychoice result. (Later build a classifier combining CX3/CX2 threshold and Myriad results).
ruben <-read_excel("/home/lythgo02/ecotyper/CNS_for_ecotyper/Copy of Supplementary Tables 58-64 Ruben CX.xlsx", 
    sheet = "ST_58_Clinical_data_TCGA-OV") %>% 
  clean_names() %>% 
  dplyr::rename(id=name)

myriad_pos <- ruben %>%  dplyr::filter(myriadmy_choice == "Positive") # HRD
myriad_neg<- ruben %>%  dplyr::filter(myriadmy_choice == "Negative")  # HRP
myriad_na<- ruben %>%  dplyr::filter(myriadmy_choice == "NA")

#pull out the cases with HRD according to myriad mychoice result (positive)

ecotype_assignment_HRD <- merge(ecotype_assignment, myriad_pos)
ecotype_assignment_HRP <- merge(ecotype_assignment, myriad_neg)
ecotype_assignment_NA <- merge(ecotype_assignment, myriad_na)

```


```{r}
#pulling out HRD and HRP from the analysis performed on the whole TCGA cohort
HRD_counts_per_ecotype <- ecotype_assignment_HRD %>% 
  group_by(assigned_ce) %>% 
  tally() 

#HRD_counts_per_ecotype <- rbind(HRD_counts_per_ecotype, newrow) 
HRD_counts_per_ecotype$n <- as.numeric(HRD_counts_per_ecotype$n) # converting back to numeric as adding the new "unassigned" row converts to character
HRD_proportions_per_ecotype <-  HRD_counts_per_ecotype %>%  adorn_percentages("col")  %>% mutate(percentage=n*100)# convert counts to percentages
  
tcga_HRD_ov <- ggplot(HRD_proportions_per_ecotype, aes(x=assigned_ce, y=percentage, fill=assigned_ce)) +
  geom_col() +
  theme_bw() + 
  xlab("Carcinoma Ecotype") + 
  ylab("%") +
  ggtitle("HRD cohort") +
  theme(legend.position = "none")+
   scale_x_discrete(limits = c("CE1", "CE2", "CE3", "CE4", "CE5", "CE6", "CE7", "CE8", "CE9", "CE10"))+
  theme(text=element_text(size=10), #change font size of all text
        axis.text=element_text(size=8), #change font size of axis text
        axis.title=element_text(size=10), #change font size of axis titles
        plot.title=element_text(size=10), #change font size of plot title
        legend.text=element_text(size=10), #change font size of legend text
        legend.title=element_text(size=10))+ #change font size of legend title  +    
  scale_color_hue(c=30, l=80) 
tcga_HRD_ov


HRP_counts_per_ecotype <- ecotype_assignment_HRP %>% 
  group_by(assigned_ce) %>% 
  tally() 

#HRD_counts_per_ecotype <- rbind(HRD_counts_per_ecotype, newrow) 
HRP_counts_per_ecotype$n <- as.numeric(HRP_counts_per_ecotype$n) # converting back to numeric as adding the new "unassigned" row converts to character
HRP_proportions_per_ecotype <-  HRP_counts_per_ecotype %>%  adorn_percentages("col")  %>% mutate(percentage=n*100)# convert counts to percentages
  
tcga_HRP_ov <- ggplot(HRP_proportions_per_ecotype, aes(x=assigned_ce, y=percentage, fill=assigned_ce)) +
  geom_col() +
  theme_bw() + 
  xlab("Carcinoma Ecotype") + 
  ylab("%") +
  ggtitle("HRP cohort") +
  theme(legend.position = "none") +
   scale_x_discrete(limits = c("CE1", "CE2", "CE3", "CE4", "CE5", "CE6", "CE7", "CE8", "CE9", "CE10"))+
  theme(text=element_text(size=10), #change font size of all text
        axis.text=element_text(size=8), #change font size of axis text
        axis.title=element_text(size=10), #change font size of axis titles
        plot.title=element_text(size=10), #change font size of plot title
        legend.text=element_text(size=10), #change font size of legend text
        legend.title=element_text(size=10))+ #change font size of legend title  +    
  scale_color_hue(c=30, l=80) 

gridExtra::grid.arrange(tcga_HRD_ov, tcga_HRP_ov)


```


```{r}
#retrieve the HRD and HRP subgroups from gene expression matrix

hugo_tcga_all_samples_060722 <- read_delim("TCGA_stratification/hugo_tcga_all_samples_060722.tsv", 
   delim = "\t", escape_double = FALSE, 
   trim_ws = TRUE)

hugo_tcga_all_samples_transpose <- sjmisc::rotate_df(hugo_tcga_all_samples_060722, rn="id")

hugo_tcga_all_samples_transpose <- hugo_tcga_all_samples_transpose %>% row_to_names(row_number = 1) 

hugo_tcga_all_samples_transpose$id <- hugo_tcga_all_samples_transpose$gene_name
hugo_tcga_all_samples_transpose <- hugo_tcga_all_samples_transpose %>% dplyr::select(-"gene_name")

hugo_tcga_all_samples_transpose$id <-  str_replace_all(hugo_tcga_all_samples_transpose$id, "\\.", "-")
hugo_tcga_all_samples_transpose$id <- str_trunc(hugo_tcga_all_samples_transpose$id, 15)
hugo_tcga_all_samples_transpose$id <- str_replace_all(hugo_tcga_all_samples_transpose$id, "\\...", "")


hugo_tcga_hrd <- dplyr::filter(hugo_tcga_all_samples_transpose,id %in% myriad_pos$id)

hugo_tcga_eco <- sjmisc::rotate_df(hugo_tcga_hrd, rn="id") 

write_delim(hugo_tcga_eco, file="/home/lythgo02/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcga_hrd_macthed_120722.tsv", delim="\t")
```


```{r}
query_hrd <- GDCquery(project = "TCGA-OV", 
                  data.category = "Transcriptome Profiling", 
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "STAR - Counts",
                  barcode =  c(myriad_pos$id))

GDCdownload(query_hrd)
hrd <-GDCprepare(query_hrd) #232 of the HRD samples have RNA seq 

sample_info_hrd <- as.data.frame(colData(hrd))


#hrd_clinical <- GDCquery(project = "TCGA-OV", 
              #data.category = "Clinical", 
             # file.type = "xml", 
              #barcode = myriad_pos$id)

#GDCdownload(hrd_clinical)

#clinical_treatment_hrd <- GDCprepare_clinic(hrd_clinical, clinical.info = "drug")
```


```{r}
library(SummarizedExperiment) 
# get expression matrix
hrd_data <- assay(hrd, "tpm_unstrand") %>% as.data.frame() %>% rownames_to_column("gene_id")

# get genes information
genes.info <- as.data.frame(rowRanges(hrd))  #table 

# get sample information
sample.info <- as.data.frame(colData(hrd))
```

```{r}
#match ensembl id with hugo gene names 
genes.info.hgnc <- genes.info %>% dplyr::select(gene_id, gene_name)

hugo_tcga_hrd <- left_join(hrd_data, genes.info.hgnc) %>% 
  group_by(gene_name) %>% 
  summarise(across(where(is.numeric),list(median = median)))

write_delim(hugo_tcga_hrd, file="/home/lythgo02/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcga_hrd_120722.tsv", delim="\t")
```


```{r}
#repeat for hrp cohort
query_hrp <- GDCquery(project = "TCGA-OV", 
                 data.category = "Transcriptome Profiling", 
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "STAR - Counts",
                  barcode =  c(myriad_neg$id))

GDCdownload(query_hrp)
hrp <-GDCprepare(query_hrp)

```
```{r}
# get expression matrix
hrp_data <- assay(hrp, "tpm_unstrand") %>% as.data.frame() %>% rownames_to_column("gene_id")

# get genes information
genes.info2 <- as.data.frame(rowRanges(hrp))  #table 

# get sample information
sample.info2 <- as.data.frame(colData(hrp))
```
```{r}
genes.info.hgnc2 <- genes.info2 %>% dplyr::select(gene_id, gene_name)

hugo_tcga_hrp <- left_join(hrp_data, genes.info.hgnc2) %>% 
  group_by(gene_name) %>% 
  summarise(across(where(is.numeric),list(median = median)))

write_delim(hugo_tcga_hrp, file="/home/lythgo02/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcga_hrp_120722.tsv", delim="\t")
```
```{r}
#pulling out HRD and HRP from the analysis performed separate HRD and HRP cohorts
hugo_tcga_hrd_120722 <- read_delim("/home/lythgo02/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcga_hrd_120722/Carcinoma_Ecotypes/Ecotype_Assignment.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE) %>% 
  clean_names() %>% 
  dplyr::rename(assigned_ce=carcinoma_ecotype)

HRD_120722_counts_per_ecotype <- hugo_tcga_hrd_120722 %>% 
  group_by(assigned_ce) %>% 
  tally() 

#HRD_counts_per_ecotype <- rbind(HRD_counts_per_ecotype, newrow) 
HRD_120722_counts_per_ecotype$n <- as.numeric(HRD_120722_counts_per_ecotype$n) # converting back to numeric as adding the new "unassigned" row converts to character
HRD_120722_proportions_per_ecotype <-  HRD_120722_counts_per_ecotype %>%  adorn_percentages("col")  %>% mutate(percentage=n*100)# convert counts to percentages
  
tcga_HRD_120722_ov <- ggplot(HRD_120722_proportions_per_ecotype, aes(x=assigned_ce, y=percentage, fill=assigned_ce)) +
  geom_col() +
  theme_bw() + 
  xlab("Carcinoma Ecotype") + 
  ylab("%") +
  ggtitle("HRD cohort") +
  theme(legend.position = "none")+
   scale_x_discrete(limits = c("CE1", "CE2", "CE3", "CE4", "CE5", "CE6", "CE7", "CE8", "CE9", "CE10"))+
  theme(text=element_text(size=10), #change font size of all text
        axis.text=element_text(size=8), #change font size of axis text
        axis.title=element_text(size=10), #change font size of axis titles
        plot.title=element_text(size=10), #change font size of plot title
        legend.text=element_text(size=10), #change font size of legend text
        legend.title=element_text(size=10))+ #change font size of legend title  +    
  scale_color_hue(c=30, l=80) 
tcga_HRD_120722_ov


hugo_tcga_hrp_120722 <- read_delim("/home/lythgo02/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcga_hrp_120722/Carcinoma_Ecotypes/Ecotype_Assignment.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE) %>% 
  clean_names() %>% 
  dplyr::rename(assigned_ce=carcinoma_ecotype)

HRP_120722_counts_per_ecotype <- hugo_tcga_hrp_120722 %>% 
  group_by(assigned_ce) %>% 
  tally() 

#HRD_counts_per_ecotype <- rbind(HRD_counts_per_ecotype, newrow) 
HRP_120722_counts_per_ecotype$n <- as.numeric(HRP_120722_counts_per_ecotype$n) # converting back to numeric as adding the new "unassigned" row converts to character
HRP_120722_proportions_per_ecotype <-  HRP_120722_counts_per_ecotype %>%  adorn_percentages("col")  %>% mutate(percentage=n*100)# convert counts to percentages
  
  tcga_HRP_120722_ov <- ggplot(HRP_120722_proportions_per_ecotype, aes(x=assigned_ce, y=percentage, fill=assigned_ce)) +
  geom_col() +
  theme_bw() + 
  xlab("Carcinoma Ecotype") + 
  ylab("%") +
  ggtitle("HRP cohort") +
  theme(legend.position = "none") +
   scale_x_discrete(limits = c("CE1", "CE2", "CE3", "CE4", "CE5", "CE6", "CE7", "CE8", "CE9", "CE10"))+
  theme(text=element_text(size=10), #change font size of all text
        axis.text=element_text(size=8), #change font size of axis text
        axis.title=element_text(size=10), #change font size of axis titles
        plot.title=element_text(size=10), #change font size of plot title
        legend.text=element_text(size=10), #change font size of legend text
        legend.title=element_text(size=10))+ #change font size of legend title  +    
  scale_color_hue(c=30, l=80) 

gridExtra::grid.arrange(tcga_HRD_120722_ov, tcga_HRP_120722_ov)


```
```{r}
high_hrd <- ruben %>%  dplyr::filter(hrd_score > 65 )

query_high_hrd <- GDCquery(project = "TCGA-OV", 
                  data.category = "Transcriptome Profiling", 
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "STAR - Counts",
                  barcode =  c(high_hrd$id))

GDCdownload(query_high_hrd)
high_hrd <-GDCprepare(query_high_hrd) #232 of the HRD samples have RNA seq 

sample_info_high_hrd <- as.data.frame(colData(high_hrd))
```


```{r}
library(SummarizedExperiment) 
# get expression matrix
high_hrd_data <- assay(high_hrd, "tpm_unstrand") %>% as.data.frame() %>% rownames_to_column("gene_id")

# get genes information
genes.info.high.hrd <- as.data.frame(rowRanges(high_hrd))  #table 

# get sample information
sample.info.high.hrd <- as.data.frame(colData(high_hrd))
```

```{r}
#match ensembl id with hugo gene names 
genes.info.hgnc.high.hrd <- genes.info.high.hrd %>% dplyr::select(gene_id, gene_name)

hugo_tcga_high_hrd <- left_join(high_hrd_data, genes.info.hgnc.high.hrd) %>% 
  group_by(gene_name) %>% 
  summarise(across(where(is.numeric),list(median = median)))

write_delim(hugo_tcga_high_hrd, file="/home/lythgo02/ecotyper/downloading_ecotyper_data/TCGA_stratification/hugo_tcga_high_hrd_120722.tsv", delim="\t")
```