---
title: "tcga_ov_ecotyper_prep"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#https://benbermanlab.com/assets/code/Workshop%20-%20TCGA%20data%20analysis.html

#http://www.ensembl.org/info/data/biomart/biomart_r_package.html

```{r}

suppressMessages({
    library(TCGAbiolinks)
    library(MultiAssayExperiment)
    library(maftools)
    library(dplyr)
    library(ComplexHeatmap)
    library(readr)
    library(biomaRt)
})
```

GDC database connection:
```{r}
clinical <- GDCquery_clinic("TCGA-OV")
head(clinical)
```

```{r}
query <- GDCquery(project = "TCGA-OV", 
                  data.category = "Transcriptome Profiling", 
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "HTSeq - FPKM",
                  barcode =  c("TCGA-10-0927-01A",
"TCGA-13-0891-01A",
"TCGA-23-1028-01A",
"TCGA-36-1577-01A",
"TCGA-04-1362-01A",
"TCGA-29-1696-01A",
"TCGA-59-A5PD-01A",
"TCGA-13-0900-01B",
"TCGA-24-1924-01A",
"TCGA-24-1425-01A",
"TCGA-24-1469-01A",
"TCGA-25-1870-01A",
"TCGA-04-1648-01A",
"TCGA-23-1114-01B",
"TCGA-61-1738-01A",
"TCGA-25-1632-01A",
"TCGA-25-1323-01A",
"TCGA-59-2348-01A",
"TCGA-13-0893-01B",
"TCGA-VG-A8LO-01A",
"TCGA-61-1741-01A",
"TCGA-57-1584-01A",
"TCGA-24-1424-01A",
"TCGA-24-2297-01A",
"TCGA-24-1417-01A",
"TCGA-57-1586-01A",
"TCGA-29-1707-02A",
"TCGA-13-1501-01A",
"TCGA-23-1023-01A",
"TCGA-24-2261-01A",
"TCGA-13-0762-01A",
"TCGA-24-2035-01A",
"TCGA-61-1911-01A",
"TCGA-25-2400-01A",
"TCGA-10-0936-01A",
"TCGA-61-1737-01A",
"TCGA-25-2396-01A",
"TCGA-23-1111-01A",
"TCGA-13-1407-01A",
"TCGA-61-1918-01A",
"TCGA-61-2104-01A",
"TCGA-61-1725-01A",
"TCGA-09-1665-01B",
"TCGA-13-0908-01B",
"TCGA-24-1553-01A",
"TCGA-25-1626-01A",
"TCGA-61-2110-01A",
"TCGA-24-2254-01A",
"TCGA-61-1728-01A",
"TCGA-59-2350-01A",
"TCGA-23-1024-01A",
"TCGA-25-1320-01A",
"TCGA-24-1847-01A",
"TCGA-24-1845-01A",
"TCGA-13-1405-01A",
"TCGA-24-1418-01A",
"TCGA-09-2051-01A",
"TCGA-29-1766-01A",
"TCGA-29-1763-01A",
"TCGA-13-1492-01A",
"TCGA-61-1721-01A",
"TCGA-09-0367-01A",
"TCGA-30-1714-01A",
"TCGA-23-1029-01B",
"TCGA-25-1322-01A",
"TCGA-24-1422-01A",
"TCGA-31-1959-01A",
"TCGA-24-2033-01A",
"TCGA-24-1435-01A",
"TCGA-29-2427-01A",
"TCGA-24-1923-01A",
"TCGA-61-2111-01A",
"TCGA-04-1332-01A",
"TCGA-24-2271-01A",
"TCGA-09-1668-01B",
"TCGA-24-1551-01A",
"TCGA-24-1565-01A",
"TCGA-09-1669-01A",
"TCGA-13-0804-01A",
"TCGA-13-0887-01A",
"TCGA-61-2102-01A",
"TCGA-24-2298-01A",
"TCGA-24-1431-01A",
"TCGA-04-1341-01A",
"TCGA-5X-AA5U-01A",
"TCGA-13-0768-01A",
"TCGA-24-1104-01A",
"TCGA-13-1489-01A",
"TCGA-10-0933-01A",
"TCGA-13-1498-01A",
"TCGA-13-0720-01A",
"TCGA-25-1635-01A",
"TCGA-29-1778-01A",
"TCGA-23-1122-01A",
"TCGA-13-1506-01A",
"TCGA-25-1628-01A",
"TCGA-25-1634-01A",
"TCGA-61-2009-01A",
"TCGA-24-2024-01A"

))
GDCdownload(query)
raw.counts <- GDCprepare(query = query, summarizedExperiment = FALSE) %>% 
    rename(ensembl_gene_id_version = X1) #rename to match ensemble biomart notation
```

```{r}
head(raw.counts)
```

```{r}
write_delim(raw.counts, file="tcga-ov-first100.tsv", delim="\t") #write into tab delimited file
```

Biomart: retrieve HUGO gene symbols
```{r}
library(biomaRt)

#To get the list of all the Ensembl mart available on the ensembl.org website
listEnsembl()

```
```{r}
ensembl = useEnsembl(biomart="ensembl") ##The "useEnsembl" function allow you to connect to a an ensembl website mart by specifying a BioMart and dataset parameters.

species <- listDatasets(ensembl)# to get a list of available species (mart datasets) 	

#to connect to Ensembl live gene mart human dataset (GRCh38)
###ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl") 

#or to create a mart of human dataset of the Ensembl GRCh37
grch37 = useEnsembl(biomart="ensembl",GRCh=37, dataset="hsapiens_gene_ensembl")

head(listFilters(grch37)) #The "listFilters" function will give you the list of available filters for a given mart and species:

head(listAttributes(grch37)) #The "listAttributes" function will give you the list of the available attributes for a given mart and species:
```
```{r}
#The "getBM" function allows you to build a BioMart query using a list of mart filters and attributes.
 genes <- getBM(attributes=c('ensembl_gene_id','hgnc_symbol','ensembl_gene_id_version'), mart = grch37)

TCGA_genes <- raw.counts[,1]# create tbl of gene id versions present in TCGA datasets 
TCGA_genes <- substr(TCGA_genes$ensembl_gene_id_version, 1,15) %>% # truncate to 15 characters to remove decimals (check all ensembl IDs are 15 characters before the decimal as standard)
      as.tibble() %>% 
      rename(ensembl_gene_id=value)

TCGA_genes_matched <- TCGA_genes %>% left_join(genes, by='ensembl_gene_id')


```


```{r}
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- raw.counts$X1
#raw.counts<-raw.counts[,-101]
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=genes,mart= mart)
merge(genes,G_list,by.x="X1",by.y="ensembl_gene_id")
```

